larpandoracontent/LArControlFlow/StitchingCosmicRayMergingTool.cc

```diff
@@ -532,206 +533,115 @@ void StitchingCosmicRayMergingTool::StitchPfos(const MasterAlgorithm *const pAlg
+                    }
+                    shiftedPfos.push_back(pPfoToShift);
                 }
             }
         }
-    }
-}
-
-//------------------------------------------------------------------------------------------------------------------------------------------
 
-const ParticleFlowObject *StitchingCosmicRayMergingTool::GetClosestPfo(const ParticleFlowObject *const pPfoToEnlarge, const PfoVector &pfoVector) const
-{
-    const ParticleFlowObject *pClosestPfo(*pfoVector.begin());
-
-    float minDistance(std::numeric_limits<float>::max());
-    for (const ParticleFlowObject *const pPfoToShift : pfoVector)
-    {
-        if (pPfoToShift == pPfoToEnlarge)
-            return pPfoToEnlarge;
-
-        const float thisDistance(LArPfoHelper::GetTwoDSeparation(pPfoToShift, pPfoToEnlarge));
-        if (thisDistance < minDistance)
+        // now merge all pfos
+        for (const ParticleFlowObject *const pPfoToDelete : shiftedPfos)
         {
-            minDistance = thisDistance;
-            pClosestPfo = pPfoToShift;
-        }
-    }
-
-    return pClosestPfo;
-}
-
-//------------------------------------------------------------------------------------------------------------------------------------------
-
-void StitchingCosmicRayMergingTool::FindStitchedLArTPCs(const PfoVector &pfoVector, const PfoToLArTPCMap &pfoToLArTPCMap,
-    LArTPCPair &stitchedLArTPCs) const
-{
-    for (const ParticleFlowObject *const pPfoToShift : pfoVector)
-    {
-        PfoToLArTPCMap::const_iterator iter(pfoToLArTPCMap.find(pPfoToShift));
-
-        if (iter != pfoToLArTPCMap.end())
-        {
-            const LArTPC *const pLArTPC(iter->second);
-
-            if (pLArTPC && (stitchedLArTPCs.first == pLArTPC || stitchedLArTPCs.second == pLArTPC))
+            if (pPfoToDelete == pPfoToEnlarge)
                 continue;
 
-            if (!stitchedLArTPCs.first)
-            {
-                stitchedLArTPCs.first = pLArTPC;
-            }
-            else if (!stitchedLArTPCs.second)
-            {
-                stitchedLArTPCs.second = pLArTPC;
-            }
-            else
-            {
-                throw StatusCodeException(STATUS_CODE_FAILURE);
-            }
-        }
-        else
-        {
-            throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+            pAlgorithm->StitchPfos(pPfoToEnlarge, pPfoToDelete, pfoToLArTPCMap);
         }
+
+        stitchedPfosToX0Map.insert(PfoToFloatMap::value_type(pPfoToEnlarge, x0));
     }
 }
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
 void StitchingCosmicRayMergingTool::CalculateX0(const PfoToLArTPCMap &pfoToLArTPCMap, const ThreeDPointingClusterMap &pointingClusterMap,
-    const PfoVector &pfoVector, float &x0, PfoToPointingVertexMap &pfoToPointingVertexMap) const
+    const PfoVector &pfoVector, float &x0, PfoToPointingVertexMatrix &pfoToPointingVertexMatrix) const
```

Comment:
At line 660 (not one of your updates), the code appears to select the second pfo relative to the first. However, this code starts by pointing to the same pfo as iter1. Presumably the CanTPCsBeStitched check prevents this potential stitch from proceeding, but it would be more efficient to use the std::next(iter1) approach alluded to previously, and skip the comparison to self altogether.

Reply:
agreed